method/stLearn/method_stLearn.py

```diff
@@ -0,0 +1,235 @@
+        Image.MAX_IMAGE_PIXELS=None
+        im = Image.open(image_tif)
+        image = np.array(im)
+        spatial_key = "spatial"
+        library_id = "sample"
+        quality = "hires"
+        try:
+            adata.uns[spatial_key] = {library_id: {}}
+            adata.uns[spatial_key][library_id]["images"] = {}
+            adata.uns[spatial_key][library_id]["images"][quality] = image
+            adata.uns[spatial_key][library_id]["use_quality"] = quality
+            # if technology == 'Visium':
+            #     adata.uns[spatial_key] = {library_id: {}}
+            #     adata.uns[spatial_key][library_id]["images"] = {}
+            #     adata.uns[spatial_key][library_id]["images"][quality] = image
+            #     adata.uns[spatial_key][library_id]["use_quality"] = quality
+            if technology != 'Visium':
+                adata.uns[spatial_key][library_id]["scalefactors"] = {}
+                if image_json is not None:
+                    with open(image_json) as json_file:
+                        scale_factors = json.load(json_file)
+                    scale = scale_factors.get('scale', 1.0)
+                    spot_diameter_fullres = scale_factors.get('spot_dimeter_fullres', 50.0)
+                else:
+                    scale = 1.0
+                    spot_diameter_fullres = 50.0
+
+                adata.uns[spatial_key][library_id]["scalefactors"]["tissue_hires_scalef"] = scale
+                adata.uns[spatial_key][library_id]["scalefactors"]["spot_diameter_fullres"] = spot_dimeter_fullres
+                adata.obsm[spatial_key] = adata.obs[["imagecol", "imagerow"]].values
+                adata.obs[["imagecol", "imagerow"]] = adata.obsm["spatial"] * scale
+
+            print("Added tissue image to the object.")
+        except:
+            raise ValueError(
+                f"""\
+                {args.image!r} does not end on a valid extension.
+                """
+                )
+        # Pre-processe image
+        tile_dir = Path(out_dir / "tiles")
+        tile_dir.mkdir(parents=True, exist_ok=True)
+        # Pre-processing for spot image
+        st.pp.tiling(adata, tile_dir)
+        # This step uses deep learning model to extract high-level features from tile images
+        # May need few minutes to be completed
+        st.pp.extract_feature(adata)
+        # Convert the NumPy array to Pandas DataFrame with row names
+        embedding_df = pd.DataFrame(adata.obsm['X_morphology'], index=adata.obs_names)
+    else:
+        adata.uns["spatial"] = None
+
+    return adata
+
+
+adata = get_anndata(args)
+
+
+# Set seed
+random.seed(seed)
+np.random.seed(seed)
+
+if args.dim_red is not None:
+    # Use the provided reduced dimensionality representation.
+    adata.obsm["X_pca"] = pd.read_table(args.dim_red, index_col=0).loc[adata.obs_names].to_numpy()
+else:
+    # Calculate and use PCA
+    st.pp.scale(adata)
+    st.em.run_pca(adata, n_comps=config["n_comps"], random_state=seed)
```

Comment:
Is this just a norml PCA?

If yes we don't need to handle this case and assume that you are always getting a reduced dimensionality passed via args

Reply:
Yes, it is just a normal PCA. Ok, I'll delete these.